integrations/mcp/src/haystack_integrations/tools/mcp/mcp_toolset.py

```diff
@@ -0,0 +1,232 @@
+        server_info=SSEServerInfo(base_url="http://some-remote-server.com:8000"),
+        tool_names=["add", "subtract"]  # Only include specific tools
+    )
+
+    # Use the toolset as shown in the pipeline example above
+    ```
+    """
+
+    def __init__(
+        self,
+        server_info: MCPServerInfo,
+        tool_names: list[str] | None = None,
+        connection_timeout: int = 30,
+        invocation_timeout: int = 30,
+    ):
+        """
+        Initialize the MCP toolset.
+
+        :param server_info: Connection information for the MCP server
+        :param tool_names: Optional list of tool names to include. If provided, only tools with
+                          matching names will be added to the toolset.
+        :param connection_timeout: Timeout in seconds for server connection
+        :param invocation_timeout: Default timeout in seconds for tool invocations
+        :raises MCPToolNotFoundError: If any of the specified tool names are not found on the server
+        """
+        # Initialize with empty tools list first
+        super().__init__(tools=[])
+
+        # Store configuration
+        self.server_info = server_info
+        self.tool_names = tool_names
+        self.connection_timeout = connection_timeout
+        self.invocation_timeout = invocation_timeout
+
+        # Connect and load tools
+        try:
+            # Create the appropriate client using the factory method
+            client = self.server_info.create_client()
+
+            # Connect and get available tools using AsyncExecutor
+            tools = AsyncExecutor.get_instance().run(client.connect(), timeout=self.connection_timeout)
+
+            # If tool_names is provided, validate that all requested tools exist
+            if self.tool_names:
+                available_tools = {tool.name for tool in tools}
+                missing_tools = set(self.tool_names) - available_tools
+                if missing_tools:
+                    message = (
+                        f"The following tools were not found: {', '.join(missing_tools)}. "
+                        f"Available tools: {', '.join(available_tools)}"
+                    )
+                    raise MCPToolNotFoundError(
+                        message=message, tool_name=next(iter(missing_tools)), available_tools=list(available_tools)
+                    )
+
+            # This is the function will give Tool instance to invoke
+            def create_invoke_tool(client, tool_name, invocation_timeout):
+                def invoke_tool(**kwargs) -> Any:
+                    """Invoke a tool using the existing client and AsyncExecutor."""
+                    result = AsyncExecutor.get_instance().run(
+                        client.call_tool(tool_name, kwargs), timeout=invocation_timeout
+                    )
+                    return result
+
+                return invoke_tool
+
+            # Create Tool instances not MCPTool for each available tool and add them
```

Comment:
Maybe also a naive question but why aren't we using MCPTool here? From an outside perspecitve it does feel a little odd that a MCPToolSet is not a Set of MCPTools. Totally fine if this is how it needs to be, just curious.

Reply:
Thanks!